trie/sync.go

```diff
@@ -122,64 +137,64 @@ func (batch *syncMemBatch) hasCode(hash common.Hash) bool {
-	nodeReqs map[common.Hash]*request // Pending requests pertaining to a trie node hash
-	codeReqs map[common.Hash]*request // Pending requests pertaining to a code hash
-	queue    *prque.Prque             // Priority queue with the pending requests
-	fetches  map[int]int              // Number of active fetches per trie node depth
+	database ethdb.KeyValueReader         // Persistent database to check for existing entries
+	membatch *syncMemBatch                // Memory buffer to avoid frequent database writes
+	nodeReqs map[string]*nodeRequest      // Pending requests pertaining to a trie node path
+	codeReqs map[common.Hash]*codeRequest // Pending requests pertaining to a code hash
+	queue    *prque.Prque                 // Priority queue with the pending requests
+	fetches  map[int]int                  // Number of active fetches per trie node depth
 }
 
 // NewSync creates a new trie data download scheduler.
 func NewSync(root common.Hash, database ethdb.KeyValueReader, callback LeafCallback) *Sync {
 	ts := &Sync{
 		database: database,
 		membatch: newSyncMemBatch(),
-		nodeReqs: make(map[common.Hash]*request),
-		codeReqs: make(map[common.Hash]*request),
+		nodeReqs: make(map[string]*nodeRequest),
+		codeReqs: make(map[common.Hash]*codeRequest),
 		queue:    prque.New(nil),
 		fetches:  make(map[int]int),
 	}
-	ts.AddSubTrie(root, nil, common.Hash{}, callback)
+	ts.AddSubTrie(root, nil, common.Hash{}, nil, callback)
 	return ts
 }
 
-// AddSubTrie registers a new trie to the sync code, rooted at the designated parent.
-func (s *Sync) AddSubTrie(root common.Hash, path []byte, parent common.Hash, callback LeafCallback) {
+// AddSubTrie registers a new trie to the sync code, rooted at the designated
+// parent for completion tracking. The given path is a unique node path in
+// hex format and contain all the parent path if it's layered trie node.
+func (s *Sync) AddSubTrie(root common.Hash, path []byte, parent common.Hash, parentPath []byte, callback LeafCallback) {
 	// Short circuit if the trie is empty or already known
 	if root == emptyRoot {
 		return
 	}
-	if s.membatch.hasNode(root) {
+	if s.membatch.hasNode(path) {
 		return
 	}
-	// If database says this is a duplicate, then at least the trie node is
-	// present, and we hold the assumption that it's NOT legacy contract code.
 	if rawdb.HasTrieNode(s.database, root) {
 		return
 	}
 	// Assemble the new sub-trie sync request
-	req := &request{
-		path:     path,
+	req := &nodeRequest{
 		hash:     root,
+		path:     path,
 		callback: callback,
 	}
 	// If this sub-trie has a designated parent, link them together
 	if parent != (common.Hash{}) {
-		ancestor := s.nodeReqs[parent]
+		ancestor := s.nodeReqs[string(parentPath)]
 		if ancestor == nil {
 			panic(fmt.Sprintf("sub-trie ancestor not found: %x", parent))
 		}
 		ancestor.deps++
-		req.parents = append(req.parents, ancestor)
+		req.parent = ancestor
```

Comment:
I think we should add a check her,e that if `req.parent != nil { panic }`

Reply:
Otherwise we'll never notice if there's some double referencing as we'd just blindly overwrite the old parent.

Reply:
req.parent can never be nil, since it's just constructed.

But I get your point. Theoretically a node should only be queued for retrieving one time, (all the retry mechanism logic should be implemented outside), maybe we can add check there?

Reply:
Nvm, I misread something